Modules/_io/textio.c

```diff
@@ -793,12 +794,197 @@ static const encodefuncentry encodefuncs[] = {
     {NULL, NULL}
 };
 
+static int
+validate_newline(const char *newline)
+{
+    if (newline && newline[0] != '\0'
+        && !(newline[0] == '\n' && newline[1] == '\0')
+        && !(newline[0] == '\r' && newline[1] == '\0')
+        && !(newline[0] == '\r' && newline[1] == '\n' && newline[2] == '\0')) {
+        PyErr_Format(PyExc_ValueError,
+                     "illegal newline value: %s", newline);
```

Comment:
Perhaps use `%r` here? If there are control characters in `newline`, it will be difficult to read otherwise.

Reply:
I can't because newline is `char*`.
And this code is not changed, just moved.